File: QUARTER_FINALS/2021/C.cpp
// #include <iostream>
// #include <math.h>
// #include <map>
// #include <set>
// #include <iomanip>
// #include <vector>
// #include <climits>
// #include <string>
// #include <algorithm>
// #include <queue>
// #include <list>
// #include <stack>

// #define ll long long

// using namespace std;

// struct Entry {
//     int i;
//     char sport;
//     ll fa, fb;
// };

// bool comp(const Entry &e1, const Entry &e2) {
//     ll diff1 = e1.fa - e1.fb;
//     ll diff2 = e2.fa - e2.fb;
//     return diff1 > diff2;
// }

// int main() {
//     ios_base::sync_with_stdio(false);
//     cin.tie(0);
//     cout.tie(0);

//     cout << fixed << setprecision(9);

//     int t;
//     cin >> t;

//     for (int _ = 0; _ < t; _++) {
//         ll n, k, x, y;
//         cin >> n >> k >> x >> y;

//         vector<Entry> guys (n);
//         for (int i = 0; i < n; i++) {
//             char type; int force;
//             cin >> type >> force;

//             guys[i].i = i;
//             guys[i].sport = type;
//             guys[i].fa = guys[i].fb = force;
//             if (type == 'A')
//                 guys[i].fb -= x;
//             else
//                 guys[i].fa -= y;
//         }

//         sort(guys.begin(), guys.end(), comp);

//         vector<ll> pref (n, LLONG_MIN);
//         ll suma = 0;
//         priority_queue<ll, vector<ll>, greater<ll>> heapa;
//         for (int i = 0; i < n; i++) {
//             suma += guys[i].fa;
//             heapa.push(guys[i].fa);

//             if (heapa.size() > k) {
//                 suma -= heapa.top();
//                 heapa.pop();
//             }
//             if (heapa.size() == k) {
//                 pref[i] = suma;
//             }
//         }

//         vector<ll> suff (n + 1, LLONG_MIN);
//         ll sumb = 0;
//         priority_queue<ll, vector<ll>, greater<ll>> heapb;
//         for (int i = n - 1; i >= 0; i--) {
//             sumb += guys[i].fb;
//             heapb.push(guys[i].fb);

//             if (heapb.size() > k) {
//                 sumb -= heapb.top();
//                 heapb.pop();
//             }
//             if (heapb.size() == k) {
//                 suff[i] = sumb;
//             }
//         }

//         ll best = LLONG_MIN;
//         for (int i = k - 1; i < n - k; i++) {
//             if (pref[i] != LLONG_MIN && suff[i + 1] != LLONG_MIN)
//                 best = max(best, pref[i] + suff[i + 1]);
//         }

//         cout << best << "\n";
//     }
// }

#include <iostream>
#include <math.h>
#include <map>
#include <set>
#include <iomanip>
#include <vector>
#include <climits>
#include <string>
#include <algorithm>
#include <queue>
#include <list>
#include <stack>
    
#define ll long long
    
using namespace std;
    
struct Entry {
    int i;
    char sport;
    ll fa, fb;
};
    
bool comp(const Entry &e1, const Entry &e2) {
    ll diff1 = e1.fa - e1.fb;
    ll diff2 = e2.fa - e2.fb;
    return diff1 > diff2;
}
    
int main() {
    // ios_base::sync_with_stdio(false);
    cin.tie(0);
    cout.tie(0);
    
    cout << fixed << setprecision(9);
    
    int t;
    cin >> t;
    
    for (int _ = 0; _ < t; _++) {
        ll n, k, x, y;
        cin >> n >> k >> x >> y;
    
        vector<Entry> guys (n);
        for (int i = 0; i < n; i++) {
            char type; int force;
            cin >> type >> force;
    
            guys[i].i = i;
            guys[i].sport = type;
            guys[i].fa = guys[i].fb = force;
            if (type == 'A')
                guys[i].fb -= x;
            else
                guys[i].fa -= y;
        }
    
        sort(guys.begin(), guys.end(), comp);
    
        vector<ll> pref (n + 1, LLONG_MIN);
        ll suma = 0;
        priority_queue<int, vector<int>, greater<int>> heapa;
        for (int i = 0; i < n; i++) {
            suma += guys[i].fa;
            heapa.push(guys[i].fa);
    
            if (heapa.size() > k) {
                suma -= heapa.top();
                heapa.pop();
            }
            if (heapa.size() == k) {
                pref[i + 1] = suma;
            }
        }
    
        vector<ll> suff (n + 1, LLONG_MIN);
        ll sumb = 0;
        priority_queue<int, vector<int>, greater<int>> heapb;
        for (int i = n - 1; i >= 0; i--) {
            sumb += guys[i].fb;
            heapb.push(guys[i].fb);
    
            if (heapb.size() > k) {
                sumb -= heapb.top();
                heapb.pop();
            }
            if (heapb.size() == k) {
                suff[i] = sumb;
            }
        }
    
        ll best = LLONG_MIN;
        for (int i = k - 1; i <= n + 1 - k; i++) {
            if (pref[i] != LLONG_MIN && suff[i] != LLONG_MIN)
                best = max(best, pref[i] + suff[i]);
        }
    
        cout << best << "\n";
    }
}

File: QUARTER_FINALS/2021/C.py
import sys
import math
input = sys.stdin.readline

INF_NEG = -10**30

t = int(input())
for _ in range(t):
    n, k, x, y = map(int, input().split())
    people = []
    for _ in range(n):
        s, f = input().split()
        f = int(f)
        if s == 'A':
            vA = f
            vB = f - x
        else:  # s == 'B'
            vA = f - y
            vB = f
        d = vA - vB
        people.append((d, vA, vB))

    # Sort by d descending
    people.sort(key=lambda z: z[0], reverse=True)

    # Prefix: best sum of top-k vA among indices [0..i]
    import heapq
    pref = [INF_NEG] * n
    heapA = []
    sumA = 0
    for i in range(n):
        a = people[i][1]
        heapq.heappush(heapA, a)
        sumA += a
        if len(heapA) > k:
            sumA -= heapq.heappop(heapA)
        if len(heapA) == k:
            pref[i] = sumA

    # Suffix: best sum of top-k vB among indices [i..n-1]
    suf = [INF_NEG] * (n + 1)  # suf[i] uses indices i..n-1
    heapB = []
    sumB = 0
    for i in range(n - 1, -1, -1):
        b = people[i][2]
        heapq.heappush(heapB, b)
        sumB += b
        if len(heapB) > k:
            sumB -= heapq.heappop(heapB)
        if len(heapB) == k:
            suf[i] = sumB

    # Try every split: prefix [0..i] for A, suffix [i+1..n-1] for B
    ans = INF_NEG
    for i in range(k - 1, n - k):
        if pref[i] != INF_NEG and suf[i + 1] != INF_NEG:
            ans = max(ans, pref[i] + suf[i + 1])

    print(ans)
